python/ray/data/datasource/databricks_uc_datasource.py

```diff
@@ -0,0 +1,188 @@
+                "schema": self.schema,
+            }
+        )
+
+        req_auth = ("token", self.token)
+        req_headers = {"Content-Type": "application/json"}
+
+        response = requests.post(
+            url_base,
+            auth=("token", self.token),
+            headers={"Content-Type": "application/json"},
+            data=payload,
+        )
+        response.raise_for_status()
+        statement_id = response.json()["statement_id"]
+
+        state = response.json()["status"]["state"]
+
+        logger.info(f"Waiting for query {query} exeuction result.")
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
+                response = requests.get(
+                    urljoin(url_base, statement_id) + "/",
+                    auth=req_auth,
+                    headers=req_headers,
+                )
+                response.raise_for_status()
+                state = response.json()["status"]["state"]
+        except KeyboardInterrupt:
+            # User cancel the command, so we cancel query execution.
+            requests.post(
+                urljoin(url_base, f"{statement_id}/cancel"),
+                auth=req_auth,
+                headers=req_headers,
+            )
+            try:
+                response.raise_for_status()
+            except Exception as e:
+                logger.warning(
+                    f"Canceling query {query} execution failed, reason: {repr(e)}."
+                )
+            raise
+
+        if state != "SUCCEEDED":
+            raise RuntimeError(f"Query {self.query} execution failed.")
+
+        manifest = response.json()["manifest"]
+        is_truncated = manifest["truncated"]
+
+        if is_truncated:
+            logger.warning(
+                f"The result dataset of '{query}' exceeding 100GiB and it is truncated."
+            )
+
+        chunks = manifest["chunks"]
+
+        # Make chunks metadata are ordered by index.
+        chunks = sorted(chunks, key=lambda x: x["chunk_index"])
+        num_chunks = len(chunks)
+        self.num_chunks = num_chunks
+        self._estimate_inmemory_data_size = sum(chunk["byte_count"] for chunk in chunks)
+
+        def get_read_task(task_index, parallelism):
+            # 0 <= task_index < parallelism
+            chunk_index_list = list(range(task_index, parallelism, num_chunks))
+
+            num_rows = sum(
+                chunks[chunk_index]["row_count"] for chunk_index in chunk_index_list
+            )
+            size_bytes = sum(
+                chunks[chunk_index]["byte_count"] for chunk_index in chunk_index_list
+            )
+
+            metadata = BlockMetadata(
+                num_rows=num_rows,
+                size_bytes=size_bytes,
+                schema=None,
+                input_files=None,
+                exec_stats=None,
+            )
+
+            def read_fn():
+                for chunk_index in chunk_index_list:
```

Comment:
Should we be reading the last chunk only after all of the other chunks have been read? IIRC if you read the last chunk, Databricks marks the statement as closed and you can't access any of the other chunks.

Reply:
uh ? I don't think we have this limitation. Databricks warehouse caches all chunks data in server side for 1 hour, and we can fetch all chunks in any order and we can fetch chunk data multiple times. Let me confirm this with our databricks warehouse engineers.

Reply:
I have confirmed with our databricks warehouse engineer, he says my understanding is correct.
 it used to work like it's described like you said, but not anymore.

Reply:
Gotcha. That's awesome